### ISSUE TITLE
FireMonkey 2.0

### ISSUE THREAD

**Issue by erosman** [erosman/support#258]
I have decided to finally use CodeMirror for for easier editing & syntax highlighting. A number of additional CodeMirror add-ons has also been added to extend editing capabilities.

While the editor will not be as streamlined as previously, it will have a lot more capabilities. The use of CodeMirror only affects script editing page. The background operation of FireMonkey (e.g. script injection & handling) which is greatly streamlined, remains the same as before.

Theme has also been implemented with 5 Light (default, eclipse, elegant, mdn-like, neat) & 5 Dark themes (cobalt, darcula, lucario, oceanic-next, rubyblue) have been added.

Please let me know if there is any thing else to consider.

**Comment by CrendKing** [comment]
Last time you mentioned in #235 I thought you gonna release soon. Since you are taking time, I'd mention some other feature requests.

1. FireMonkey supports both CSS and script. I usually have both custom CSS and script for a specific website, and their names used to be e.g. "Google Search - Add favicon to each entry" or "Google Search - Custom font CSS". This habit was carried from back when I was using Greasemonkey and Stylus. Soon I realized the problem: since FireMonkey sort the items alphabetically, CSS and script would mix with each other, and when I want to change or disable some of them (usually the scripts, because Google changed their DOM names again...) it becomes harder to find. So now I name everything with prefix like "CSS: Google Search - custom font", "Script: Google search blah blah blah". This way all CSS appear first, then all scripts. It works, but other users might share the same (frustrating) experience. So I wonder if you want to do something about it, such as put them in different folders respectively (e.g. a CSS folder, a script folder, a tree structure with two root nodes).
2. Back when I use Stylus, every time I change the CSS, it refreshes the used CSS on all injected tabs, so that I can observe the effect without reloading the page. FireMonkey does not do this. Since it relies on Firefox to do the injection, I don't know if you can mimic the behavior. But if you can, can you consider? If you concern about performance, put it behind an option maybe?

**Comment by erosman** [comment]
>  Since you are taking time, ....

It is done and I am testing & finalising. I had to make some changes to CodeMirror, some of which has been approved by CodeMirror and will be in their next release, and some customisation that are specific to FireMonkey. There are also new tools and features.
I was also looking forward to comments and/or requests before releasing v2.0.
I have also been trying to find a way to make it compatible with **sync** `GM_getValue` but haven't found any way to do it.

> 1. FireMonkey supports both CSS and script .....

I do the same if I have CSS & JS for the same site e.g. site & site CSS.
In both options page and toolbar popup, the icon next to it makes it easy for me to differentiate.
In options page there is also the selection filter on top to show JS or CSS etc.

I haven't had any issues. Lets hear from other users and decide.

> 2. Back when I use Stylus, every time I change the CSS,

Stylus, Stylish, etc as well as GM, TM, VM manually inject script or CSS by checking the URL against the script or CSS parameters and if there is a match, inject it.

FireMonkey passes it all to Firefox at start up and everything is handled natively by Firefox. Therefore, there is a considerable efficiency.

The problem is that the injected script or CSS can not be removed.

Firefox has a CSS removable API (that Stylus, Stylish, etc use) but that is only for manual use.

However, FireMonkey has Scratchpad which can inject CSS and remove it (v1.45) since it uses the above API.

Personally, I use **F12 -> Developer Tools -> Style Editor** and live-edit the style that was injected until I get the desired outcome, then copy it back to the UserCSS.

**Comment by CrendKing** [comment]
> In options page there is also the selection filter on top to show JS or CSS etc.

But the popup has no filter, and that's where the most trouble came, plus it is a much smaller area to work with.

> The problem is that the injected script or CSS can not be removed.

Understandable. I'll use scratchpad or style editor then.

**Comment by erosman** [comment]
> But the popup has no filter

hm ... it is possible to add the filter there as well .. let me think about it.

**Comment by CrendKing** [comment]
Sure. Just remember since I've already done all the renaming, it won't affect me anymore. You might need to make some judgment call.

**Comment by erosman** [comment]
> Just remember since I've already done all the renaming, it won't affect me anymore. You might need to make some judgment call.

Let's see if there is more demand for it.

**Comment by erosman** [comment]
`@antifeatures` is added to v2.0 and waiting for it to be finalised. (ref: https://github.com/JasonBarnabe/greasyfork/issues/604).

**Comment by erosman** [comment]
v2.0 released

**Comment by CrendKing** [comment]
Thanks. Here's some immediate problems I encountered. If you prefer a separate issue, I can delete the comment and transfer.

1. I can't create any new script or CSS. Click +, immediately save. Nothing happens. Click save again, it prompts me "already exists". Click OK to replace, then refresh the page. That name is nowhere to be found. And if I repeat the process, the first time it does not prompt me "already exists". However, save actually works.
2. Probably related to 1, I can't update the template either.
3. What is this? Doesn't do anything. A bug?
![Report](https://user-images.githubusercontent.com/975235/101538670-4d2f9e80-3952-11eb-8eff-dd1f9955ac18.jpg)
4. Desperately need some functionality to customize the theme. For example, dracula hard code Consolas font. Other themes use my default monospace font, but much smaller. Also for instance I prefer light theme but I always use gray background color instead of pure white.
5. Also need functionality to customize editor rules. For example, I use 4 indentation spaces, but currently it hard codes to 2.
6. Lots of false positive lint errors. For example, some section lines (`includeGlobs`) are error. Javascript missing semicolon is error (I strongly reject the useless semicolon in JS). `URLSearchParams` is "not defined". `location` is read only.

**Comment by erosman** [comment]
> If you prefer a separate issue,

No problem.. here is fine.

> 1. I can't create any new script or CSS.
I will test and get back to you.

> 2. Probably related to 1, I can't update the template either.
I will test and get back to you.

> 3. What is this? Doesn't do anything. A bug?
Display any script. If it has lint errors, you can use the REPORT to see all them.

> 4. Desperately need some functionality to customize the theme.
I can work on that in future.
You can also try [CodeMirror Theme Demo](https://codemirror.net/demo/theme.html) to see if there are any good ones to include. I tried to include a few ones.

> 5. Also need functionality to customize editor rules.
I can work on that in future but there are limits to what can be changed.

> 6. Lots of false positive lint errors. 
Most of them are generated by [JSHint](https://github.com/jshint/jshint) and a very few are added by FireMonkey. I have set it to be more strict than other script-managers. However, the lint errors are only guides and information. You can ignore them.

**Comment by CrendKing** [comment]
> I can work on that in future.

Actually if you can allow to apply a CSS to the editor, that would be enough. ViolentMonkey has that feature.

> there are limits on what can be changed

https://codemirror.net/doc/manual.html#config

**Comment by erosman** [comment]
> ViolentMonkey has that feature.

![image](https://user-images.githubusercontent.com/6458876/101590576-f6809e00-39ff-11eb-98cd-16cabf597092.png)

It seems the feature is for the overall options page.
GM & TM don't have such options.

> https://codemirror.net/doc/manual.html#config

![image](https://user-images.githubusercontent.com/6458876/101590743-43fd0b00-3a00-11eb-8d7a-db1d7924755c.png)

The CodeMirror in FireMonkey is customised, and some settings can break it. I will see what is possible.
GM & TM don't have such options.

**Comment by erosman** [comment]
> 1. I can't create any new script or CSS.

Fixed

> 2. Probably related to 1, I can't update the template either.

Fixed

**Comment by CrendKing** [comment]
> It seems the feature is for the overall options page.

You can try this yourself:
```
textarea {
    font-family: serif !important;
}

.CodeMirror {
    font-family: monospace !important;
}
```
`textarea` applies only to the textareas in the settings page. `.CodeMirror` applies to the editor.

> I will see what is possible.

I'm not asking for full list. Just the bare minimum, like `indentUnit`.

**Comment by erosman** [comment]
> textarea applies only to the textareas in the settings page. .CodeMirror applies to the editor. 

I will work on it.

> I'm not asking for full list. Just the bare minimum, like indentUnit.

OK... I will add the feature.

**Comment by erosman** [comment]
> 6. Lots of false positive lint errors. For example, some section lines (includeGlobs) are error. Javascript missing semicolon is error (I strongly reject the useless semicolon in JS). URLSearchParams is "not defined". location is read only. And apparently this line is wrong? grid-template-columns: repeat(3, 1fr);

Lint control options added v2.01

**Comment by CrendKing** [comment]
Thanks. I got v2.3. Two problems:

1. It seems the custom CSS is not working. I don't see it injected anywhere. For example, I put in
```
.CodeMirror {
    font-family: monospace !important;
}

textarea {
    font-family: serif !important;
}
```
But nothing's changed.

2. I can confirm the CodeMirror settings JSON is working, though it is very limited at the moment. Looking forward to have more settings in the future 👍 The problem is I can't seem to get the lint option working. I tried
```
{
  "indentUnit": 4,
  "lint": {
    "javascript": {
      "lastsemic": false
    }
  }
}
```
but once I click save it removes everything inside "lint". Can you provide an example?

**Comment by erosman** [comment]
> 1. It seems the custom CSS is not working. I don't see it injected anywhere. For example, I put in

It is injected in the last `<style></style>` in the `<head>` e.g.

```html
    <!-- after CodeMirror to override -->
    <link rel="stylesheet" href="cm.css">
    <!-- custom style -->
    <style></style>
  </head>
  <body>
```

It is injected after all other CSS so there shouldn't be a need for `!important` as it would override previous ones (unless they also have `!important`). I will check it now again.

> 2. ....  The problem is I can't seem to get the lint option working. I tried

Check the **Help**... it has a list and method. Here is a copy;

### Allowed Configurations & their Current Values
```json
{
  "indentWithTabs": false,
  "indentUnit": 2,
  "tabSize": 2,
  "lint": {
    "curly": true,
    "devel": true,
    "eqeqeq": true,
    "freeze": true,
    "latedef": "nofunc",
    "leanswitch": true,
    "maxerr": 100,
    "noarg": true,
    "nonbsp": true,
    "undef": true,
    "unused": true,
    "varstmt": true
  }
}
```

**Comment by CrendKing** [comment]
> I will check it now again.

Just verified from inspector that the CSS is not there:
```
    <!-- custom style -->
    <style></style>
```

> Check the **Help**... it has a list and method

Ah, I see it. I tried just putting `"lastsemic": false` there but not working. I don't see any relaxing option. Guess you are trying to making the lint strict, which is usually good. But in my case if every line has a exclamation icon on it due to "missing semicolon", the real useful warnings are overshadowed. So for people who don't use semicolon, the lint is completely useless. Of course I can live with it, so no worry.

**Comment by erosman** [comment]
> I tried just putting "lastsemic": false there but not working. I don't see any relaxing option. Guess you are trying to making the lint strict

I see... no.. since I had not set it myself, it was not included in the allowed parameters. Let me see CSSLint options.
If there are options, I may have to change the format slightly.

**Comment by CrendKing** [comment]
May I ask what prevents you from pass the JSON directly to the underlying linter? Why do you have to explicitly allow each option?

**Comment by erosman** [comment]
I have fixed the issues with Custom CSS.
I have update the format of CodeMirror options

```
{
  "indentWithTabs": false,
  "indentUnit": 2,
  "tabSize": 2,
  "jshint": {
    // any option except: globals & jquery
  }
}
```


> May I ask what prevents you from pass the JSON directly to the underlying linter as-is? Why do you have to explicitly allow each option?

There are some specific ones that will break the CodeMirror. There are some that are specifically added by FireMonkey for it to functions. FIreMonkey switches between CSS & JavaScript each time and some settigns relate to those switches.
There are also a few that will be in next version of CodeMirror (I had made a PR which were done).

I guess, I can allow all and just disallow the few that cause problems!! ... like:

```
theme               // it is set from the menu
mode                // that breaks everything
lint                // messes with switching between modes
search              // will be in the next CodeMirror
extraKeys           // maybe add something later.. for now it will clear the keyboard shortcuts
```

**Comment by erosman** [comment]
v2.4 is out.. check the **Help** and note the new format for e.g.`" jshint"`

**Comment by CrendKing** [comment]
Perfect! I got everything I've every wanted for a script manager 👍 Thank you!

**Comment by benoit-pierre** [comment]
I miss the option to have a simple text edit widget, no syntax: with that, I could externally edit my scripts/themes with a proper editor, and automatically update the text in the Firemonkey options tab through the accessibility API (using a small Python script) on save. Now I have to wrangle with cut/copy/paste...

**Comment by erosman** [comment]
> I miss the option to have a simple text edit widget, no syntax: with that,

If there is a popular demand for it, I will see if I can work something out.

> I could externally edit my scripts/themes with a proper editor,

I normally do the same... and then drag & drop on Firefox tab to update the script (or keep the script open in the tab and refresh, when I do a lot of edits).

**Comment by CrendKing** [comment]
> through the accessibility API

Didn't know that's possible. Always did copy paste if I use external editor. Is there any detail of the Python script you could share? And why the new design prevents you from using that anymore?

**Comment by benoit-pierre** [comment]
This is the [script](https://gist.github.com/benoit-pierre/330482b5c10f9514914046cfac060789), I would:
* open the Firemonkey options in a new Firefox window, select the script I want to edit
* run the script like this `EDITOR='gvim -f' firemonkey_external_edit.py my.script.user.js`
* from now on, every time the script is saved in the editor, the copy in the Firemonkey editor is updated (and the save button clicked)

Hide the Firemonkey window, and I could just alternate between a Firefox window with the page I'm testing the script on and my editor.

> Also, if any external program could access Firefox content through accessibility API, isn't that big security hole? Always thought accessibility is a backdoor and should be disabled unless you can't use Firefox without it.

¯\\_(ツ)_/¯ That's the extremes you have to go to in the fabulous world of a WebExtensions...

**Comment by CrendKing** [comment]
I see. So basically you are doing something equivalent to DOM manipulation outside of Firefox. And because there is no more textarea, you actually need to use [CodeMirror programming API](https://codemirror.net/doc/manual.html#setValue) to update the content, which I suppose the accessibility does not provide.

> ¯\_(ツ)_/¯ That's the extremes you have to go to in the fabulous world of a WebExtensions...

I understand your frustration, but I don't think it has anything to do with WebExtension. What you are doing is breaching the content sandbox of the browser, not the extension. If you can write to a textarea in a webpage, you can certainly read a password textbox in a webpage too, and that applies to any application in the system.

I think what erosman should actually do, if he decides to support it, is build an application with [native messaging](https://www.google.com/search?q=firefox+native+api) that communicates to FireMonkey. Then he can link that application with the EDITOR so that any save triggers content update and save. This way only this specific application can access a specific webextension in Firefox so that you can disable accessibility API for security and prevent unauthorized 3rd party access.

**Comment by erosman** [comment]
You might be able to do a workaround... for now.

The` textarea` is still there, that is not a problem. 

The problem is that on **SAVE**, FM gets _CodeMirror_ to damp content back to the textarea so it will wipe out your changes.

However .......... the 2 tools "_Tabs to Spaces_" & "_Trim Trailing spaces_" work directly with the `textarea` content and then refresh the _CodeMirror_ content.

Therefore, you can:
- set content to the `textarea`
- trigger "_Trim Trailing spaces_"
- **save** whenever you are happy with the result

It is a workaround ... but workable.

**Comment by CrendKing** [comment]
@erosman There is a bug. I just write it down here. Feel free to move to separate issue if appropriate.

You know you can disable individual item in the popup. If I disable some items, then go to editor and save any item (even without any change), all the items go back to be enabled. However, if I manually disable the item in the editor (top-right menu), it remains disabled.

**Comment by erosman** [comment]
> You know you can disable individual item in the popup. If I disable some items, then go to editor and save any item (even without any change), all the items go back to be enabled. However, if I manually disable an item in the editor (top-right menu) then save any, it remains disabled.

Oops ... there was an oversight when changed to ES6 modules (getting enable & auto update input values). 
Fixed for v2.6
